libethereum/EthereumHost.cpp

```diff
@@ -512,26 +514,30 @@ tuple<vector<shared_ptr<EthereumPeer>>, vector<shared_ptr<EthereumPeer>>, vector
     vector<shared_ptr<EthereumPeer>> chosen;
     vector<shared_ptr<EthereumPeer>> allowed;
     vector<shared_ptr<SessionFace>> sessions;
-
-    size_t peerCount = 0;
+    double percentDecimal = _percent / 100.0;
     foreachPeer([&](std::shared_ptr<EthereumPeer> _p)
     {
         if (_allow(_p.get()))
         {
-            allowed.push_back(_p);
             sessions.push_back(_p->session());
+            allowed.push_back(_p);
         }
-        ++peerCount;
         return true;
     });
 
-    size_t chosenSize = (peerCount * _percent + 99) / 100;
+    if(_percent == 0 || allowed.size() == 0)
+    {
+        return make_tuple(move(chosen), move(allowed), move(sessions));
+    }
+
+    std::shuffle(allowed.begin(), allowed.end(), *m_urng);
+
+    size_t chosenSize = percentDecimal * allowed.size();
     chosen.reserve(chosenSize);
-    for (unsigned i = chosenSize; i && allowed.size(); i--)
+    for(size_t i = allowed.size() - 1; i >= allowed.size() - chosenSize; i--)
     {
-        unsigned n = rand() % allowed.size();
-        chosen.push_back(std::move(allowed[n]));
-        allowed.erase(allowed.begin() + n);
+        chosen.push_back(move(allowed[i]));
+        allowed.erase(allowed.end() - i);
```

Comment:
this loop body was supposed to move an element from `allowed` to `chosen`. You seem to first move value of `allowed[i]` but then erase `allowed[size - i]`, doesn't look correct

This whole loop is confusing, maybe just create new vectors like
```
vector<shared_ptr<EthereumPeer>> chosen(allowed.begin(), allowed.begin() + chosenSize);
vector<shared_ptr<EthereumPeer>> rest(allowed.begin() + chosenSize, allowed.end());
```
?